# ML-07 — Baseline Action Score and Top-20 Review

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/AbdullahIdrees291/flyrank-ml-internship/blob/main/work/notebooks/w04_baseline_score.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My rule and its reason codes

*Write the rule in plain words first. Then the reason codes it can output.*

In [19]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [20]:
import pandas as pd
import os

print("Current folder:")
print(os.getcwd())

print("\nFiles/folders:")
print(os.listdir())

Current folder:
/content/flyrank-ml-internship

Files/folders:
['work', 'LICENSE', 'docs', '02_your_first_readable_model.ipynb', 'w03_data_contract.ipynb', '.gitignore', '.github', 'README.md', 'skills', '.git', 'w02_ml_task_framing.ipynb', 'outputs', 'AGENTS.md', '01_first_look_and_discovery.ipynb', 'GUIDE.md', 'submission', 'SETUP.md', 'CLAUDE.md', 'requirements.txt', 'DATA_USE.md', 'data', 'scripts', 'notebooks']


In [21]:
# Check what data and variables are already available
print("Available variables:")
print([x for x in globals().keys() if not x.startswith("_")])

# Show DataFrames that are already loaded
for name, obj in list(globals().items()):
    if hasattr(obj, "shape") and hasattr(obj, "columns"):
        print("\nDATAFRAME:", name)
        print("Shape:", obj.shape)
        print("Columns:")
        print(list(obj.columns))

Available variables:
['In', 'Out', 'get_ipython', 'exit', 'quit', 'pd', 'os', 'name', 'obj', 'root', 'dirs', 'files', 'file', 'DATA_PATH', 'df', 'np', 'work', 'stale', 'stale_bins', 'stale_labels', 'stale_check', 'volume', 'volume_bins', 'volume_labels', 'volume_check', 'queue', 'output_cols', 'baseline_queue', 'output_path', 'top20', 'review_cols', 'top20_review', 'weak_picks', 'rule_inputs', 'forbidden_terms', 'used_columns', 'possible_leakage']

DATAFRAME: df
Shape: (30000, 44)
Columns:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_

In [22]:
!git clone https://github.com/AbdullahIdrees291/flyrank-ml-internship.git

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 139, done.
remote: Counting objects: 100% (139/139), done.
remote: Compressing objects: 100% (94/94), done.
remote: Total 139 (delta 49), reused 95 (delta 29), pack-reused 0 (from 0)
Receiving objects: 100% (139/139), 1.86 MiB | 9.19 MiB/s, done.
Resolving deltas: 100% (49/49), done.


In [23]:
import os

os.chdir("/content/flyrank-ml-internship")

print("Current folder:", os.getcwd())
print(os.listdir())

Current folder: /content/flyrank-ml-internship
['work', 'LICENSE', 'docs', '02_your_first_readable_model.ipynb', 'flyrank-ml-internship', 'w03_data_contract.ipynb', '.gitignore', '.github', 'README.md', 'skills', '.git', 'w02_ml_task_framing.ipynb', 'outputs', 'AGENTS.md', '01_first_look_and_discovery.ipynb', 'GUIDE.md', 'submission', 'SETUP.md', 'CLAUDE.md', 'requirements.txt', 'DATA_USE.md', 'data', 'scripts', 'notebooks']


In [24]:
import os

print("DATA FILES:")
for root, dirs, files in os.walk("data"):
    for file in files:
        print(os.path.join(root, file))

DATA FILES:
data/raw/content_refresh_anonymized.csv


In [25]:
import pandas as pd

DATA_PATH = "data/raw/content_refresh_anonymized.csv"

df = pd.read_csv(DATA_PATH)

print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

print("\nFirst 5 rows:")
display(df.head())

Shape: (30000, 44)

Columns:
['content_id', 'client_id', 'search_volume', 'competition', 'competition_level', 'cpc', 'content_type', 'main_intent', 'word_count', 'char_count', 'provider_used', 'model_used', 'impressions_90d', 'clicks_90d', 'pageviews_90d', 'sessions_90d', 'users_90d', 'engaged_sessions_90d', 'ai_sessions_90d', 'scroll_events_90d', 'days_with_impressions', 'days_with_sessions', 'impressions_last_30d', 'clicks_last_30d', 'sessions_last_30d', 'impressions_prev_30d', 'clicks_prev_30d', 'sessions_prev_30d', 'content_age_days', 'age_tier', 'age_tier_order', 'days_since_last_update', 'freshness_tier', 'word_count_tier', 'char_count_tier', 'ctr', 'avg_position', 'engagement_rate', 'scroll_rate', 'ai_traffic_pct', 'impression_tier', 'position_tier', 'trend_direction', 'trend_pct']

First 5 rows:


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


### Rule

I will prioritize content that is both stale and has meaningful search volume.

The rule uses two observable signals:
- `days_since_last_update`: higher values indicate more stale content and support a refresh action.
- `search_volume`: higher values indicate greater search demand and support prioritizing the item.

The baseline score will give more points to content that is older and has higher search volume. Higher scores mean higher priority for review/action.

### Reason codes

- `STALE_HIGH_VOLUME`: the content is relatively stale and has relatively high search volume.
- `STALE`: the content is relatively stale but does not have high search volume.
- `HIGH_VOLUME`: the content has relatively high search volume but is not especially stale.
- `LOW_PRIORITY`: neither signal is high enough to justify a strong action.

### Action labels

- `REFRESH`: stale and high-volume content.
- `REVIEW`: one of the two signals is high.
- `MONITOR`: neither signal is high.

In [26]:
# Section 1: signal checks and rule setup

import pandas as pd
import numpy as np

# Work from the already-loaded dataset
work = df.copy()

# ---------------------------------------------------------
# Signal 1: days_since_last_update
# ---------------------------------------------------------

stale = work["days_since_last_update"].dropna()

stale_bins = [-np.inf, 30, 90, 180, np.inf]
stale_labels = ["0-30", "31-90", "91-180", "181+"]

work["staleness_bucket"] = pd.cut(
    work["days_since_last_update"],
    bins=stale_bins,
    labels=stale_labels
)

stale_check = (
    work.groupby("staleness_bucket", observed=False)
        .agg(
            n=("content_id", "size"),
            median_trend_pct=("trend_pct", "median"),
            mean_trend_pct=("trend_pct", "mean")
        )
        .reset_index()
)

print("SIGNAL 1 — days_since_last_update")
display(stale_check)

# ---------------------------------------------------------
# Signal 2: search_volume
# ---------------------------------------------------------

volume = work["search_volume"].dropna()

volume_bins = [-np.inf, 10, 100, 1000, np.inf]
volume_labels = ["0-10", "11-100", "101-1000", "1001+"]

work["volume_bucket"] = pd.cut(
    work["search_volume"],
    bins=volume_bins,
    labels=volume_labels
)

volume_check = (
    work.groupby("volume_bucket", observed=False)
        .agg(
            n=("content_id", "size"),
            median_trend_pct=("trend_pct", "median"),
            mean_trend_pct=("trend_pct", "mean")
        )
        .reset_index()
)

print("SIGNAL 2 — search_volume")
display(volume_check)

# ---------------------------------------------------------
# Verdicts
# ---------------------------------------------------------

print("VERDICTS")
print("days_since_last_update: CONFIRMED if the more-stale buckets show a meaningfully different/directionally worse trend than the freshest bucket.")
print("search_volume: CONFIRMED if higher-volume buckets show a meaningful directional difference in trend or opportunity.")

SIGNAL 1 — days_since_last_update


,staleness_bucket,n,median_trend_pct,mean_trend_pct
0,0-30,20480,-33.3,0.784405
1,31-90,175,-38.5,-7.373054
2,91-180,9171,-34.4,-15.683224
3,181+,174,-44.5,-6.781203


SIGNAL 2 — search_volume


,volume_bucket,n,median_trend_pct,mean_trend_pct
0,0-10,18392,-36.20,-9.812087
1,11-100,6091,-29.25,3.514769
2,101-1000,2489,-26.00,1.736780
3,1001+,560,-18.10,20.232621


VERDICTS
days_since_last_update: CONFIRMED if the more-stale buckets show a meaningfully different/directionally worse trend than the freshest bucket.
search_volume: CONFIRMED if higher-volume buckets show a meaningful directional difference in trend or opportunity.


### Signal verdicts

- **days_since_last_update — CONFIRMED:** The measured buckets show a directional relationship: the freshest bucket had mean trend_pct of +0.78%, while older buckets were negative, including -15.68% for 91–180 days. The 181+ bucket was also negative at -6.78%. This supports using staleness as a refresh-priority signal, while recognizing that the relationship is not perfectly monotonic.

- **search_volume — CONFIRMED:** Higher search-volume buckets showed a stronger directional opportunity. Mean trend_pct was -9.81% for 0–10 volume, +1.74% for 101–1000, and +20.23% for 1001+. This supports using search volume as a prioritization signal.

## 2. Build the ranked queue (writes the CSV)

*Code the score, rank everything, write work/outputs/baseline_action_score.csv.*

In [27]:
import os

os.makedirs("work/outputs", exist_ok=True)

output_path = "work/outputs/baseline_action_score.csv"
baseline_queue.to_csv(output_path, index=False)

print("CSV written to:", output_path)
print("Rows:", len(baseline_queue))

print("\nTop 10:")
display(baseline_queue.head(10))

CSV written to: work/outputs/baseline_action_score.csv
Rows: 30000

Top 10:


,rank,content_id,score,reason_code,action,days_since_last_update,search_volume
0,1,content_bbca724138f2,6.0,STALE_HIGH_VOLUME,REFRESH,236,1600.0
1,2,content_ef99c4abd9ab,5.0,STALE_HIGH_VOLUME,REFRESH,104,74000.0
2,3,content_bf67a444faef,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
3,4,content_5ec29ae79c60,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
4,5,content_deb54e9e19cd,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
5,6,content_454cc6654c6e,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
6,7,content_7868341d97dd,5.0,STALE_HIGH_VOLUME,REFRESH,104,40500.0
7,8,content_84fe9d0a707a,5.0,STALE_HIGH_VOLUME,REFRESH,104,40500.0
8,9,content_6b41450ae50c,5.0,STALE_HIGH_VOLUME,REFRESH,104,27100.0
9,10,content_e7eb94e121b9,5.0,STALE_HIGH_VOLUME,REFRESH,104,22200.0


In [28]:
import os
print(os.getcwd())

/content/flyrank-ml-internship


In [29]:
# Section 2: Build the ranked baseline action queue

# Start from the working dataframe
queue = work.copy()

# ---------------------------------------------------------
# 1. Create simple, observable signal scores
# ---------------------------------------------------------

# Staleness score
queue["stale_score"] = pd.cut(
    queue["days_since_last_update"],
    bins=[-np.inf, 30, 90, 180, np.inf],
    labels=[0, 1, 2, 3]
).astype(float).fillna(0)

# Search-volume score
queue["volume_score"] = pd.cut(
    queue["search_volume"],
    bins=[-np.inf, 10, 100, 1000, np.inf],
    labels=[0, 1, 2, 3]
).astype(float).fillna(0)

# ---------------------------------------------------------
# 2. Combined baseline score
# ---------------------------------------------------------

queue["score"] = queue["stale_score"] + queue["volume_score"]

# ---------------------------------------------------------
# 3. ONE reason code per row
# ---------------------------------------------------------

queue["reason_code"] = np.select(
    [
        (queue["stale_score"] >= 2) & (queue["volume_score"] >= 2),
        (queue["stale_score"] >= 2),
        (queue["volume_score"] >= 2)
    ],
    [
        "STALE_HIGH_VOLUME",
        "STALE",
        "HIGH_VOLUME"
    ],
    default="LOW_PRIORITY"
)

# ---------------------------------------------------------
# 4. Action label
# ---------------------------------------------------------

queue["action"] = np.select(
    [
        queue["reason_code"] == "STALE_HIGH_VOLUME",
        queue["reason_code"].isin(["STALE", "HIGH_VOLUME"])
    ],
    [
        "REFRESH",
        "REVIEW"
    ],
    default="MONITOR"
)

# ---------------------------------------------------------
# 5. Rank the queue
# ---------------------------------------------------------

queue = queue.sort_values(
    ["score", "search_volume", "days_since_last_update"],
    ascending=[False, False, False]
).reset_index(drop=True)

queue["rank"] = queue.index + 1

# ---------------------------------------------------------
# 6. Select output columns
# ---------------------------------------------------------

output_cols = [
    "rank",
    "content_id",
    "score",
    "reason_code",
    "action",
    "days_since_last_update",
    "search_volume"
]

baseline_queue = queue[output_cols].copy()

# ---------------------------------------------------------
# 7. Write the required CSV
# ---------------------------------------------------------

output_path = "work/outputs/baseline_action_score.csv"

baseline_queue.to_csv(output_path, index=False)

print("CSV written to:", output_path)
print("Rows:", len(baseline_queue))

print("\nTop 10:")
display(baseline_queue.head(10))

CSV written to: work/outputs/baseline_action_score.csv
Rows: 30000

Top 10:


,rank,content_id,score,reason_code,action,days_since_last_update,search_volume
0,1,content_bbca724138f2,6.0,STALE_HIGH_VOLUME,REFRESH,236,1600.0
1,2,content_ef99c4abd9ab,5.0,STALE_HIGH_VOLUME,REFRESH,104,74000.0
2,3,content_bf67a444faef,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
3,4,content_5ec29ae79c60,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
4,5,content_deb54e9e19cd,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
5,6,content_454cc6654c6e,5.0,STALE_HIGH_VOLUME,REFRESH,104,60500.0
6,7,content_7868341d97dd,5.0,STALE_HIGH_VOLUME,REFRESH,104,40500.0
7,8,content_84fe9d0a707a,5.0,STALE_HIGH_VOLUME,REFRESH,104,40500.0
8,9,content_6b41450ae50c,5.0,STALE_HIGH_VOLUME,REFRESH,104,27100.0
9,10,content_e7eb94e121b9,5.0,STALE_HIGH_VOLUME,REFRESH,104,22200.0


In [30]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Top-20 review

*For each of the top 20: action, reason code, confidence note, and what would make it wrong.*

In [31]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [32]:
# Section 3: Top-20 review

top20 = baseline_queue.head(20).copy()

# Add simple confidence notes based on the baseline score
top20["confidence_note"] = np.select(
    [
        top20["score"] >= 5,
        top20["score"] >= 4,
        top20["score"] >= 2
    ],
    [
        "High: both signals are strongly elevated.",
        "Medium-high: both signals contribute meaningfully.",
        "Medium: one signal is elevated."
    ],
    default="Low: neither signal is strongly elevated."
)

# Add a skeptical "what would make it wrong" note
top20["what_would_make_it_wrong"] = np.select(
    [
        top20["reason_code"] == "STALE_HIGH_VOLUME",
        top20["reason_code"] == "STALE",
        top20["reason_code"] == "HIGH_VOLUME"
    ],
    [
        "It would be wrong if the page is intentionally old or already scheduled for an update.",
        "It would be wrong if the content is still performing well despite being old.",
        "It would be wrong if high search volume does not translate into a meaningful opportunity for this content.",
    ],
    default="It would be wrong if there is no meaningful reason to prioritize the content."
)

# Display the required top-20 review
review_cols = [
    "rank",
    "content_id",
    "action",
    "reason_code",
    "confidence_note",
    "what_would_make_it_wrong"
]

top20_review = top20[review_cols].copy()

display(top20_review)

,rank,content_id,action,reason_code,confidence_note,what_would_make_it_wrong
0,1,content_bbca724138f2,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
1,2,content_ef99c4abd9ab,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
2,3,content_bf67a444faef,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
3,4,content_5ec29ae79c60,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
4,5,content_deb54e9e19cd,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
5,6,content_454cc6654c6e,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
6,7,content_7868341d97dd,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
7,8,content_84fe9d0a707a,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
8,9,content_6b41450ae50c,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...
9,10,content_e7eb94e121b9,REFRESH,STALE_HIGH_VOLUME,High: both signals are strongly elevated.,It would be wrong if the page is intentionally...


## 4. Weak picks + leakage check

*Which picks look wrong and why? Confirm no product flags or future windows leaked in.*

In [33]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


In [34]:
# Section 4: Weak picks + leakage check

# ---------------------------------------------------------
# 1. Identify weak/questionable picks
# ---------------------------------------------------------

weak_picks = baseline_queue[
    (baseline_queue["score"] <= 2)
].head(10).copy()

print("WEAK / QUESTIONABLE PICKS")
display(weak_picks)

# ---------------------------------------------------------
# 2. Leakage check
# ---------------------------------------------------------

rule_inputs = [
    "days_since_last_update",
    "search_volume"
]

print("\nRULE INPUTS:")
print(rule_inputs)

# Check that the scoring rule only depends on the two chosen signals
print("\nLEAKAGE CHECK")
print("The baseline score uses only:")
print("- days_since_last_update")
print("- search_volume")

print("\nNo future-window or label-derived fields are used in the score.")

# Explicitly check that future/label-like fields are not used
forbidden_terms = [
    "label",
    "target",
    "outcome",
    "future",
    "next",
    "conversion"
]

used_columns = set(["days_since_last_update", "search_volume"])

possible_leakage = [
    col for col in used_columns
    if any(term in col.lower() for term in forbidden_terms)
]

print("\nPossible leakage fields among rule inputs:", possible_leakage)

if len(possible_leakage) == 0:
    print("LEAKAGE CHECK: PASS")
else:
    print("LEAKAGE CHECK: REVIEW NEEDED")

# ---------------------------------------------------------
# 3. Weak-pick explanation
# ---------------------------------------------------------

print("\nWEAK PICK REVIEW")
print(
    "Weak picks are lower-scoring rows that may not deserve immediate action. "
    "They are useful as a skeptical check because the baseline is intentionally simple."
)

WEAK / QUESTIONABLE PICKS


,rank,content_id,score,reason_code,action,days_since_last_update,search_volume
3293,3294,content_389fd04f2a91,2.0,HIGH_VOLUME,REVIEW,28,1000.0
3294,3295,content_976d5deeab73,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3295,3296,content_9d45a4710a9a,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3296,3297,content_a7452d4e5f24,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3297,3298,content_2f99e0fa3453,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3298,3299,content_dde90787c795,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3299,3300,content_0ac51e0dae75,2.0,HIGH_VOLUME,REVIEW,25,1000.0
3300,3301,content_5cae1270b78a,2.0,HIGH_VOLUME,REVIEW,22,1000.0
3301,3302,content_30e786775ba6,2.0,HIGH_VOLUME,REVIEW,22,1000.0
3302,3303,content_aa6577303b25,2.0,HIGH_VOLUME,REVIEW,22,1000.0



RULE INPUTS:
['days_since_last_update', 'search_volume']

LEAKAGE CHECK
The baseline score uses only:
- days_since_last_update
- search_volume

No future-window or label-derived fields are used in the score.

Possible leakage fields among rule inputs: []
LEAKAGE CHECK: PASS

WEAK PICK REVIEW
Weak picks are lower-scoring rows that may not deserve immediate action. They are useful as a skeptical check because the baseline is intentionally simple.


### Weak picks

The weakest picks are useful as a skeptical check. A low baseline score does not prove that an item has no opportunity; it only means that the two selected signals did not strongly prioritize it.

### Leakage check

The baseline score uses only `days_since_last_update` and `search_volume`. These are observable signals available for the rule. No future-window field, target/label, or outcome-derived field is used in the scoring calculation.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.